rasa/shared/core/events.py

```diff
@@ -317,15 +318,23 @@ def as_dict(self) -> Dict[Text, Any]:
 
         return d
 
+    def fingerprint(self) -> Text:
```

Comment:
does it make sense to replace / combine this with the existing `__hash__` implementation which we have on every event anyway?

Reply:
Hmm, I steered away from anything using python `hash()` because it's unstable, hence I don't think it's a good idea to use this.

Reply:
yep, but our custom `__hash__` implementations in the events, are stable, aren't they? (they should probably not be called `__hash__` in the first place

Reply:
Not sure, looking at [`UserUttered`](https://github.com/RasaHQ/rasa/blob/071d2ef24760f738590971c044c703cc27efb567/rasa/shared/core/events.py#L474) and `SlotSet` events for example, their custom `__hash__` uses the built-in `hash()` 🤔  While a few other events like `AllSlotsReset` have [static hashes](https://github.com/RasaHQ/rasa/blob/071d2ef24760f738590971c044c703cc27efb567/rasa/shared/core/events.py#L1046).

I'm also confused because the parent class doesn't have a `__hash__` method.

Reply:
What a mess 🙈 The neat thing is that the current `__hash__` implementation excludes all properties which aren't relevant which makes me think of one edge case: Don't get `Event`s assigned timestamps when being read? And wouldn't they differ depending on when you read the training data with the events?

Reply:
> Don't get Events assigned timestamps when being read? And wouldn't they differ depending on when you read the training data with the events?

Yes, if timestamp is not specified, then it will get the current timestamp. Although not sure I understand the 2nd question, when would the events be re-instantiated? I am probably not seeing the full implications of where events get used.

However if timestamp is a problem, I can remove it from the fingerprint in the interface.
I still wouldn't use the __hash__ implementation, but if yu think it's a must, I can also spend some effort to create custom fingerprint methods for certain events?

Reply:
Yup I just tested using `__hash__` in the fingerprint and it fails the consistency test.

Reply:
We could potentially unify `__hash__` and `fingerprint`. I.e. make `__hash__` use your new stable implementation.

Reply:
By that do you mean delete current `__hash__` body in each individual subclass and return instead `self.fingerprint()`?
Or maybe make `__hash__` use `deep_container_fingerprint()` instead of builtin `hash`?

Reply:
> By that do you mean delete current __hash__ body in each individual subclass and return instead self.fingerprint()?

Yup exactly. They are both attempting to the do the same thing, i.e. a unique representation.

For reference:
The `__hash__` method is used when you pass the object into the `hash` function and also, in turn, for things like using the object as a key in a `dictionary`, or putting it in a `set`.

Reply:
> though not sure I understand the 2nd question, when would the events be re-instantiated? I am probably not seeing the full implications of where events get used.

Sorry, I'm a bit short on time, so I hope phrasing this as question works 🙈 
If you read a story file and create a tracker from the events and then do the same thing a couple of seconds later:
Is the fingerprint of both trackers the same?

Reply:
@wochinge I'll write a test for that 😄

Reply:
@wochinge Yup, timestamp does affect the fingerprint of both the trackers. I'll keep timestamp out of the event fingerprint. And also push this new test too.

Reply:
@joejuzl regarding the __hash__ implementation: would it be ok to add __hash__ to parent class and then remove the __hash__ for each individual subclass? There would be too many duplicate return lines if I amend this in every subclass 😄

Reply:
ah in the end seems I have to go down this route of duplication otherwise I get a `TypeError: unhashable type` when trying to use hash() on specific event subclasses.

Reply:
> @joejuzl regarding the __hash__ implementation: would it be ok to add __hash__ to parent class and then remove the __hash__ for each individual subclass? There would be too many duplicate return lines if I amend this in every subclass 😄

Yes definitely.

> I get a TypeError: unhashable type when trying to use hash() on specific event subclasses.

huh that's weird. But `fingerprint` returns a `str` right?  which should also always be hashable. Worth looking into a but further I think.

Reply:
I think I found the answer [here](https://docs.python.org/3/reference/datamodel.html#object.__hash__) which is identical to Events scenario:
```
If a class that overrides __eq__() needs to retain the implementation of __hash__() from a parent class, the interpreter must be told this explicitly by setting __hash__ = <ParentClass>.__hash__.
```
So either way I still have a lot of lines to duplicate - which one do you think makes most sense? The above one or having a __hash__ implementation per subclass?